src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/shim.cxx

```diff
@@ -1211,16 +1223,55 @@ namespace xclcpuemhal2 {
       return;
     }
 
+    setDeviceProcessStarted(false);
     std::string socketName = sock->get_name();
-    if(socketName.empty() == false)// device is active if socketName is non-empty
+    // device is active if socketName is non-empty
+    if (!socketName.empty())
     {
 #ifndef _WINDOWS
       xclClose_RPC_CALL(xclClose,this);
 #endif
     }
+   closeMessengerThread();
    saveDeviceProcessOutput();
   }
 
+  void CpuemShim::closeMessengerThread()
+  {
+    if (mMessengerThread.joinable())
+      mMessengerThread.join();
+  }
+
+  void CpuemShim::messagesThread()
+  {
+    auto start_time = std::chrono::high_resolution_clock::now();
+    auto lpath = this->deviceDirectory + "/../../../device_process.log";
+    sParseLog deviceProcessLog(lpath);
+    int count = 0;
+    while (getDeviceProcessStarted())
+    {
+      // I may not get ParseLog() all the times, So Let's optimize myself.
+      // Let's sleep for 10,20,30 seconds....etc until it is < 300 seconds
+      // After that I may not get extensive parseLog() statements.
+      // So I want to call it for each interval of 300 seconds
+
+      auto end_time = std::chrono::high_resolution_clock::now();
+
+      if (std::chrono::duration_cast<std::chrono::seconds>(end_time - start_time).count() <= 300) {  
+        deviceProcessLog.parseLog(); 
+        ++count;
+        if (count%5 == 0) {
+          std::this_thread::sleep_for(std::chrono::seconds(10*(count/5)));
+          }
+        }
+
+      if (std::chrono::duration_cast<std::chrono::seconds>(end_time - start_time).count() >= 300) {  
+        std::this_thread::sleep_for(std::chrono::seconds(300));
+        deviceProcessLog.parseLog();
```

Comment:
It is almost a copy paste of 10 lines above.
I wonder whether what you are trying to do (well I am not sure what is the incantation here) cannot be simplified in some
```c++
std::this_thread::sleep_for(std::chrono::seconds(std::min(10*(count/5), 300));
```

Reply:
done.